pkg/server/sotw/v2/server.go

```diff
@@ -302,88 +210,16 @@ func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest
 				}
 			}
 
-			// cancel existing watches to (re-)request a newer version
-			switch {
-			case req.TypeUrl == resource.EndpointType:
-				if values.endpointNonce == "" || values.endpointNonce == nonce {
-					if values.endpointCancel != nil {
-						values.endpointCancel()
-					}
-					values.endpoints, values.endpointCancel = s.cache.CreateWatch(req)
-				}
-			case req.TypeUrl == resource.ClusterType:
-				if values.clusterNonce == "" || values.clusterNonce == nonce {
-					if values.clusterCancel != nil {
-						values.clusterCancel()
-					}
-					values.clusters, values.clusterCancel = s.cache.CreateWatch(req)
-				}
-			case req.TypeUrl == resource.RouteType:
-				if values.routeNonce == "" || values.routeNonce == nonce {
-					if values.routeCancel != nil {
-						values.routeCancel()
-					}
-					values.routes, values.routeCancel = s.cache.CreateWatch(req)
-				}
-			case req.TypeUrl == resource.ListenerType:
-				if values.listenerNonce == "" || values.listenerNonce == nonce {
-					if values.listenerCancel != nil {
-						values.listenerCancel()
-					}
-					values.listeners, values.listenerCancel = s.cache.CreateWatch(req)
+			typeUrl := req.TypeUrl
+			responseNonce, seen := values.nonces[typeUrl]
+			if !seen || responseNonce == nonce {
+				if cancel, seen := values.cancellations[typeUrl]; seen && cancel != nil {
+					cancel()
```

Comment:
@kyessenov To actually make guarantees about available space in channel - we can drain it here, after calling cancel. But it doesn't seem like a clean solution :(

Reply:
After thinking more about this and trying to implement draining - it does not look like a dirty hack anymore.
It allows us to actually get back to similar guarantees that we had in current implementation. And it allows to have buffer size exactly equal to number of types that will go through this channel.